### PR TITLE
native dpnp.tril

### PR DESCRIPTION
**Pull request by Rubtsowa**

### CODE REVIEW

**Review by shssf — APPROVED** [review]

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 1]
```diff
@@ -72,6 +72,85 @@ void dpnp_full_c(void* array_in, void* result, const size_t size)
...
+            if (ids[res_ndim - 1] <= diag_idx)
+            {
+                result[i] = array_m[i];
+            }
+            else
+            {
+                result[i] = 0;
+            }
+        }
+    }
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:150]: Code duplication, can be redesigned.

File: dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx [hunk 2]
```diff
@@ -54,6 +54,9 @@ __all__ += [
 ]
 
 
+ctypedef void(*custom_1in_1out_func_ptr_t)(void * , void * , const int , size_t * , size_t * , const size_t, const size_t)
```
Comment (samir-nasibli) [inline dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx:57]: 1in 1out func? seems it doesn't matches with function signature. Bad naming practice.

File: tests/test_arraycreation.py [hunk 3]
```diff
@@ -110,3 +110,29 @@ def test_loadtxt(type):
         dpnp_res = dpnp.loadtxt(fh, dtype=type)
 
         numpy.testing.assert_array_equal(dpnp_res, np_res)
+
+
+@pytest.mark.parametrize("k",
+                         [-6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6],
+                         ids=['-6', '-5', '-4', '-3', '-2', '-1', '0', '1', '2', '3', '4', '5', '6'])
```
Comment (samir-nasibli) [inline tests/test_arraycreation.py:117]: do you really need to test all these cases?
Isn't one instance enough for cases `k>0`, `k=0`, `k<0` for testing?

File: tests/test_arraycreation.py [hunk 4]
```diff
@@ -110,3 +110,29 @@ def test_loadtxt(type):
...
+                          [[1, 2], [3, 4]],
+                          [[0, 1, 2], [3, 4, 5], [6, 7, 8]],
+                          [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9]]],
+                         ids=['[0, 1, 2, 3, 4]',
+                              '[1, 1, 1, 1, 1]',
+                              '[[0, 0], [0, 0]]',
+                              '[[1, 2], [1, 2]]',
+                              '[[1, 2], [3, 4]]',
+                              '[[0, 1, 2], [3, 4, 5], [6, 7, 8]]',
+                              '[[0, 1, 2, 3, 4], [5, 6, 7, 8, 9]]'])
```
Comment (samir-nasibli) [inline tests/test_arraycreation.py:132]: the same comment, see prev. one.

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 5]
```diff
@@ -72,6 +72,85 @@ void dpnp_full_c(void* array_in, void* result, const size_t size)
     dpnp_initval_c<_DataType>(result, array_in, size);
 }
 
+template <typename _DataType>
+void dpnp_tril_c(void* array_in, void* result1, const int k, size_t* shape, size_t* res_shape, const size_t ndim, const size_t res_ndim)
+{
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:77]: there is no check for `nullptrs`.

**Review by samir-nasibli — COMMENTED** [review]
Seems to refactoring and redesign required (in other PR(s)).

File: dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx [hunk 6]
```diff
@@ -237,35 +240,24 @@ cpdef dparray dpnp_tri(N, M, k, dtype):
...
+cpdef dparray dpnp_tril(dparray m, int k):
     if m.ndim == 1:
-        result = dparray(shape=(m.shape[0], m.shape[0]), dtype=m.dtype)
+        result_shape = (m.shape[0], m.shape[0])
+    else:
+        result_shape = m.shape
 
-        for i in range(result.size):
-            ids = get_axis_indeces(i, result.shape)
+    result_ndim = len(result_shape)
```
Comment (samir-nasibli) [inline dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx:249]: unused.

File: dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx [hunk 7]
```diff
@@ -237,35 +240,24 @@ cpdef dparray dpnp_tri(N, M, k, dtype):
...
+    cdef dparray result = dparray(result_shape, dtype=m.dtype)
```
Comment (samir-nasibli) [inline dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx:250]: Maybe some typo. maybe you meant `result_ndim`:
```python
    cdef dparray result = dparray(result_ndim, dtype=m.dtype)
```

**Pull request merged**